File: js/object.js
function getData() {

	// throw Error("Failed to get data.");
	return 1;
}

class Value {

	#number

	constructor(number) {

		this.#number = number;
	}

	add(value) {

		if (value instanceof Value) {

			return new Value(this.#number + value.#number);
		}
		else {

			return this;
		}
	}

	divide(value) {

		if (value instanceof Value) {

			return new Value(this.#number / value.#number);
		}
		else {

			return this;
		}
	}

	print() {

		console.log(this.#number);
	}
}

class ValueError {

	#message

	constructor(message) {

		this.#message = message;
	}

	add() {

		return this;
	}

	divide() {

		return this;
	}

	print() {

		console.error(this.#message);
	}
}

function Data() {

	try {

		return new Value(getData());
	}
	catch (error) {

		return new ValueError(error.message);
	}
}

function Addition(left, right) {

	return left.add(right);
}

function Division(left, right) {

	return left.divide(right);
}

new Division(new Addition(new Data(), new Value(2)), new Value(2)).print();

File: kotlin/Object.kt
interface Value
{
	fun add(value: Value): Value;

	fun divide(value: Value): Value;

	fun print();
}

class ErrorValue(val message: String?): Value
{
	override fun add(value: Value): Value
	{
		return this;
	}

	override fun divide(value: Value): Value
	{
		return this;
	}

	override fun print()
	{
		System.err.println(message);
	}
}

class DoubleValue(val number: Double): Value
{
	override fun add(value: Value): Value
	{
		if (value is DoubleValue)
		{
			return DoubleValue(number + value.number);
		}
		else
		{
			return value;
		}
	}

	override fun divide(value: Value): Value
	{
		if (value is DoubleValue)
		{
			return DoubleValue(number / value.number);
		}
		else
		{
			return value;
		}
	}

	override fun print()
	{
		println(number);
	}
}

fun getData(): Double
{
	// throw Exception("Failed to get data.");
	return 1.0;
}

class Data(): Value by (try { DoubleValue(getData()); } catch(e: Exception) { ErrorValue(e.message); })

class Addition(left: Value, right: Value): Value by (left.add(right))

class Division(left: Value, right: Value): Value by (left.divide(right))

fun main()
{
	Division(Addition(Data(), DoubleValue(2.0)), DoubleValue(2.0)).print();
}
